### PR TITLE
[error] [lang] Improved error messages for illegal slicing or indexing to ti.field

### PR DESCRIPTION
**Pull request by jim19930609**
Related issue = #4620 

illegal behaviors to ti.field:
```
import taichi as ti
ti.init()
f_2d = ti.field(ti.u8, shape=(640, 480))

# illegal indexing
f_2d[0]

# illegal slicing
f_2d[0,4:5]
```

illegal indexing:
image

illegal slicing:
image

### CODE REVIEW

File: tests/python/test_numpy.py [hunk 1]
```diff
@@ -187,6 +187,13 @@ def test_index_mismatch():
         val[0, 0] = 1
 
 
+@test_utils.test()
+def test_invalid_slicing():
```
Comment (ailzhang) [inline tests/python/test_numpy.py:191]: `test_numpy.py` isn't the best place for this test :P maybe `test_field.py`?

File: python/taichi/lang/field.py [hunk 2]
```diff
@@ -289,7 +295,16 @@ def __setitem__(self, key, value):
...
     def __getitem__(self, key):
         self._initialize_host_accessors()
-        return self.host_accessors[0].getter(*self._pad_key(key))
+        # Check for potential slicing behaviour
+        # for instance: x[0, :]
+        padded_key = self._pad_key(key)
+        for key in padded_key:
+            if not isinstance(key, int):
+                raise TypeError(
+                    f"Detected ilegal element of type: {type(key)}. "
```
Comment (ailzhang) [inline python/taichi/lang/field.py:304]: typo: illegal ;)

File: python/taichi/lang/field.py [hunk 3]
```diff
@@ -205,7 +205,13 @@ def _pad_key(self, key):
             key = ()
         if not isinstance(key, (tuple, list)):
             key = (key, )
-        assert len(key) == len(self.shape)
+
+        if len(key) != len(self.shape):
+            raise AssertionError(
+                "Slicing an array or matrix from ti.field is not supported so far."
```
Comment (ailzhang) [inline python/taichi/lang/field.py:211]: Shall we simply say Slicing is not supported on ti.field? I'm a bit worried users confuse "matrix" here with the matrix element inside a ti.field ;)

File: tests/python/test_numpy.py (hunk 1 above)
Comment (ailzhang) [reply tests/python/test_numpy.py:191]: also maybe good to match the error message as well to make sure it fails at a place that we expected ;)

**Review by ailzhang — APPROVED** [review]
Thanks for fixing it!

File: tests/python/test_numpy.py (hunk 1 above)
Comment (jim19930609) [reply tests/python/test_numpy.py]: Good point, thx!

**Review by ailzhang — APPROVED** [review]
Thanks!

**Pull request merged**